### PR TITLE
Add profile names and `exec`

### PR DESCRIPTION
**Pull request by layus**
This may be a bit naive, any feedback welcome.

### CODE REVIEW

**Comment by StephenBrown2** [comment]
Some docs with an example here: https://github.com/emersion/kanshi/blob/master/kanshi.5.scd would be great also.

**Comment by layus** [comment]
On my way. Sadly it was not written on the previous attempt... Nothing to reuse ;-)

**Comment by layus** [comment]
> Some docs with an example here: [/kanshi.5.scd@`master`](https://github.com/emersion/kanshi/blob/master/kanshi.5.scd) would be great also.

@StephenBrown2 Done!

File: include/config.h [hunk 1]
```diff
@@ -29,10 +29,16 @@ struct kanshi_profile_output {
 	enum wl_output_transform transform;
 };
 
+struct kanshi_profile_command {
+	struct wl_list link;
+	char* command;
```
Comment (emersion) [inline include/config.h]: Style: `*` on the right

File: main.c [hunk 2]
```diff
@@ -115,6 +115,18 @@ static struct kanshi_mode *match_mode(struct kanshi_head *head,
...
+static void execute_profile_commands(struct kanshi_profile *profile) {
+	struct kanshi_profile_command *command;
+	wl_list_for_each(command, &profile->commands, link) {
+		int res = system(command->command);
+		if (res != 0) {
+			fprintf(stderr, "Command '%s' returned with exit value %d.\n",
+				command->command, res);
+		}
+	}
+	return;
```
Comment (emersion) [inline main.c]: Unnecessary `return`

File: kanshi.5.scd [hunk 3]
```diff
@@ -35,6 +35,33 @@ Directives are followed by space-separated arguments. Arguments can be quoted
...
+	}
+```
+
+	Note that some extra care must be taken with outputs identified by an
+	output description as the real name may change:
+
+```
+	{
+		output "Hewlett Packard LA2405 CZ40022966" mode 1920x1200
+		exec bash -c 'swaymsg workspace 1,  move workspace to output $(swaymsg -p -t get_outputs | grep "Hewlett Packard LA2405 CZ40022966" | cut -f2 -d" ")'
```
Comment (emersion) [inline kanshi.5.scd]: No need for `bash -c` here, we already run the command in a shell

File: main.c [hunk 4]
```diff
@@ -115,6 +115,18 @@ static struct kanshi_mode *match_mode(struct kanshi_head *head,
 	return NULL;
 }
 
+static void execute_profile_commands(struct kanshi_profile *profile) {
+	struct kanshi_profile_command *command;
+	wl_list_for_each(command, &profile->commands, link) {
+		int res = system(command->command);
```
Comment (emersion) [inline main.c]: Using `system` here will wait until the command has finished. This is different from what Sway does (executing them in parallel).

Blocking seems a little dangerous because it's not integrated with the Wayland event loop and could potentially lead to a crash because the Wayland socket becomes full. The alternative would be to manually `fork` and `execl`.

File: parser.c [hunk 5]
```diff
@@ -338,13 +356,31 @@ static struct kanshi_profile_output *parse_profile_output(
...
 }
 
+static struct kanshi_profile_command *parse_profile_command(
+		struct kanshi_parser *parser) {
+	// Skip the 'exec' directive.
+	if (!parser_expect_token(parser, KANSHI_TOKEN_STR)) {
+		return NULL;
+	}
+
+	if (!parser_read_line(parser) || parser->tok_str_len <= 0) {
```
Comment (emersion) [inline parser.c]: If `parser->tok_str_len <= 0`, we should probably error out.

**Review by emersion — CHANGES_REQUESTED** [review]
Thanks for your pull request! This looks pretty good to me, here are a few comments.

**Comment by layus** [comment]
@emersion Thanks for the review.
I implemented the asynchronous execution, but this ended up being mingled with my 'configurations names' change. I bundled everything instead of trying to separate them.

File: include/config.h [hunk 6]
```diff
@@ -31,6 +31,7 @@ struct kanshi_profile_output {
 
 struct kanshi_profile {
 	struct wl_list link;
+	char* name;
```
Comment (emersion) [inline include/config.h]: Style: `*` on the right

File: parser.c [hunk 7]
```diff
@@ -338,14 +338,36 @@ static struct kanshi_profile_output *parse_profile_output(
 	}
 }
 
-static struct kanshi_profile *parse_profile(struct kanshi_parser *parser) {
-	if (!parser_expect_token(parser, KANSHI_TOKEN_LBRACKET)) {
-		return NULL;
+static bool kanshi_parse_profile_name(struct kanshi_parser *parser, char** name) {
```
Comment (emersion) [inline parser.c]: Isn't `parser_read_str` enough?

Comment (layus) [reply parser.c]: Not too sure about this one, but it seems that using `parser_read_str` would requires a configuration name because `parser_read_str` will not accept an opening `{`. The way I implemented it the name remains optional and the file format backward compatible

File: include/config.h (hunk 6 above)
Comment (layus) [reply include/config.h]: Fixed in the next commit. I can rebase if you really prefer.

File: parser.c (hunk 7 above)
Comment (emersion) [reply parser.c]: I believe `parser_read_str` would just return fill `parser->tok_str` with an empty string, returning immediately, if there is no profile name?

File: include/config.h (hunk 6 above)
Comment (emersion) [reply include/config.h]: Yeah I'd prefer to keep commits separate and logically split

File: kanshi.5.scd [hunk 8]
```diff
@@ -35,6 +35,37 @@ Directives are followed by space-separated arguments. Arguments can be quoted
 	On *sway*(1), output names and descriptions can be obtained via
 	*swaymsg -t get_outputs*.
 
+*exec* <command>
+	An exec directive executes a command when the profile was successfully
+	applied. This can be used to update the compositor state to the profile
+	when not done automatically.
+
+	Commands are executed executed asynchronously and their order may not be
```
Comment (MichaelAquilina) [inline kanshi.5.scd]: typo in the doc: `executed` is written twice here

File: parser.c (hunk 7 above)
Comment (layus) [reply parser.c]: > I believe `parser_read_str` would just return fill `parser->tok_str` with an empty string, returning immediately, if there is no profile name?

It works indeed. The not-so-nice catch is that you have to reset `tok_str_len` manually. See updated patch.

File: kanshi.5.scd (hunk 8 above)
Comment (layus) [reply kanshi.5.scd]: Tahnks !

**Comment by layus** [comment]
@emersion Updated with respect to all the comments. The patches are now crystal clear ;-)

File: parser.c [hunk 9]
```diff
@@ -339,13 +339,36 @@ static struct kanshi_profile_output *parse_profile_output(
...
-	wl_list_init(&profile->outputs);
+	// Use the bracket position to generate a default profile name
+	if (profile->name == NULL) {
+		char generated_name[100];
+		int ret = snprintf(generated_name, 100,
+				"<anonymous at line %d, col %d>", parser->line, parser->col);
+		if (ret >= 0 && ret < 100) {
+			profile->name = strdup(generated_name);
+		} else {
+			profile->name = "<anonymous>";
```
Comment (emersion) [inline parser.c]: This would need to be `strdup`'ed. But honestly, I just wouldn't care

File: parser.c [hunk 10]
```diff
@@ -339,13 +339,36 @@ static struct kanshi_profile_output *parse_profile_output(
...
 	}
 
-	struct kanshi_profile *profile = calloc(1, sizeof(*profile));
-	wl_list_init(&profile->outputs);
+	// Use the bracket position to generate a default profile name
+	if (profile->name == NULL) {
+		char generated_name[100];
+		int ret = snprintf(generated_name, 100,
+				"<anonymous at line %d, col %d>", parser->line, parser->col);
+		if (ret >= 0 && ret < 100) {
```
Comment (emersion) [inline parser.c]: Ditto (but see comment below)

File: parser.c [hunk 11]
```diff
@@ -339,13 +339,36 @@ static struct kanshi_profile_output *parse_profile_output(
...
 	if (!parser_expect_token(parser, KANSHI_TOKEN_LBRACKET)) {
 		return NULL;
 	}
 
-	struct kanshi_profile *profile = calloc(1, sizeof(*profile));
-	wl_list_init(&profile->outputs);
+	// Use the bracket position to generate a default profile name
+	if (profile->name == NULL) {
+		char generated_name[100];
+		int ret = snprintf(generated_name, 100,
```
Comment (emersion) [inline parser.c]: Would be nicer to use `sizeof(generated_name)` for the max length.

File: parser.c [hunk 12]
```diff
@@ -339,13 +339,36 @@ static struct kanshi_profile_output *parse_profile_output(
...
+	wl_list_init(&profile->outputs);
+
+	// First parse an optional profile name
+	parser->tok_str_len = 0;
+	if (!parser_read_str(parser)) {
+		fprintf(stderr, "expected new profile, got %s\n",
+			token_type_str(parser->tok_type));
+		return NULL;
+	}
+	profile->name = (parser->tok_str_len == 0)? NULL: strdup(parser->tok_str);
```
Comment (emersion) [inline parser.c]: Style nit: space before `?` and `:`

File: kanshi.5.scd [hunk 13]
```diff
@@ -6,18 +6,18 @@ kanshi - configuration file
...
-A kanshi configuration file is a list of profiles. Each profile is delimited by
-brackets (*{* and *}*) and contains directives.
+A kanshi configuration file is a list of profiles. Each profile starts with an
+optional name and contains directives delimited by brackets (*{* and *}*).
 
 Example:
 
 ```
-{
+external {
```
Comment (emersion) [inline kanshi.5.scd]: Maybe keep this one without a name, to demonstrate the fact that it's optional?

**Review by emersion — CHANGES_REQUESTED** [review]
Almost there for the first patch, good work!

File: main.c [hunk 14]
```diff
@@ -71,10 +74,60 @@ static struct kanshi_profile *match(struct kanshi_state *state,
...
+	if (pid < 0) {
+		perror("Impossible to fork a new process");
+	}
+}
+
+static void execute_profile_commands(struct kanshi_profile *profile) {
+	struct kanshi_profile_command *command;
+	wl_list_for_each(command, &profile->commands, link) {
+		fprintf(stderr, "Running command '%s'\n", command->command);
+		fflush(stderr);
```
Comment (emersion) [inline main.c]: This is unnecessary, `stderr` is flushed on newline.

File: main.c [hunk 15]
```diff
@@ -71,10 +74,60 @@ static struct kanshi_profile *match(struct kanshi_state *state,
 }
 
 
+static void exec_command(char *cmd) {
+	pid_t pid, child;
+	// Fork process
```
Comment (emersion) [inline main.c]: This comment could be removed, it doesn't provide more information than the line below

File: main.c [hunk 16]
```diff
@@ -71,10 +74,60 @@ static struct kanshi_profile *match(struct kanshi_state *state,
...
+	if ((pid = fork()) == 0) {
+		setsid();
+		sigset_t set;
+		sigemptyset(&set);
+		sigprocmask(SIG_SETMASK, &set, NULL);
```
Comment (emersion) [inline main.c:84]: Why is this necessary? It makes it impossible to kill the process with e.g. SIGINT, right?

**Review by emersion — CHANGES_REQUESTED** [review]
Second patch looks pretty good too

File: parser.c (hunk 9 above)
Comment (layus) [reply parser.c]: It is just useless as profiles are never freed. But I can certainly do that, just in case someone implements config reload and tries to free the right pointers.

File: parser.c (hunk 10 above)
Comment (layus) [reply parser.c]: Seems the order is not preserved. I guess you mean the comment about `sizeof(generated_name)` ?

File: main.c (hunk 16 above)
Comment (layus) [reply main.c:84]: Well, this is a quick copy paste from sway's `exec` implementation, so you probably know better than me what this does. I will ave to look into that code again.

Comment (layus) [reply main.c:84]: What these three lines do is clear (reset) the signal mask, so that no signal is blocked anymore, if any was blocked before. See https://man.openbsd.org/sigsetmask.3#EXAMPLES. This clears any special mask that could have been used by kanshi, or that could be used later, for example SIGUSR1, as is the case in sway. (See https://github.com/swaywm/sway/commit/7d8413d9628ed493790454410a28b9f0c41444e6)

It think it belongs here.

**Comment by layus** [comment]
@emersion Updated and answered comments.

File: kanshi.5.scd [hunk 17]
```diff
@@ -35,6 +35,37 @@ Directives are followed by space-separated arguments. Arguments can be quoted
...
+	}
+```
+
+	Note that some extra care must be taken with outputs identified by an
+	output description as the real name may change:
+
+```
+	complex {
+		output "Hewlett Packard LA2405 CZ40022966" mode 1920x1200
+		exec swaymsg workspace 1, move workspace to output $(swaymsg -p -t get_outputs | grep "Hewlett Packard LA2405 CZ40022966" | cut -f2 -d" ")
```
Comment (emersion) [inline kanshi.5.scd]: `move workspace to output "Hewlett Packard LA2405 CZ40022966"` should just work

File: main.c [hunk 18]
```diff
@@ -71,10 +74,60 @@ static struct kanshi_profile *match(struct kanshi_state *state,
...
+	pid_t pid, child;
+	if ((pid = fork()) == 0) {
+		setsid();
+		sigset_t set;
+		sigemptyset(&set);
+		sigprocmask(SIG_SETMASK, &set, NULL);
+		if ((child = fork()) == 0) {
+			execl("/bin/sh", "/bin/sh", "-c", cmd, (void *)NULL);
+			fprintf(stderr, "Executing cmd '%s' failed: ", cmd);
+			perror("");
```
Comment (emersion) [inline main.c]: It would be better to just `fprintf(stderr, "Executing cmd '%s' failed: %s", cmd, strerror(errno));`

File: main.c [hunk 19]
```diff
@@ -71,10 +74,60 @@ static struct kanshi_profile *match(struct kanshi_state *state,
...
+			exit(-1);
+		}
+		if (child < 0) {
+			fprintf(stderr, "Impossible to fork a new process to execute command '%s': ", cmd);
+			perror("");
```
Comment (emersion) [inline main.c]: Ditto

**Review by emersion — CHANGES_REQUESTED** [review]
Okay, let's merge this after these two minor comments are fixed :)

File: parser.c (hunk 12 above)
Comment (emersion) [reply parser.c]: Bump

File: parser.c (hunk 10 above)
Comment (emersion) [reply parser.c]: Yes, `ret < sizeof(generated_name)`

File: parser.c (hunk 9 above)
Comment (emersion) [reply parser.c]: Yeah

**Comment by emersion** [comment]
Ah, missed that the first patch still has some unresolved comments

File: kanshi.5.scd (hunk 17 above)
Comment (layus) [reply kanshi.5.scd]: no, it does not. But testing further shows that `move workspace to output '"Hewlett Packard LA2405 CZ40022966"'` does.

Comment (emersion) [reply kanshi.5.scd]: Oh, yeah, i3 legacy…

Comment (layus) [reply kanshi.5.scd]: no, it does not. But testing further shows that `move workspace to output '"Hewlett Packard LA2405 CZ40022966"'` does. Changing to that, but it is potentially a bug in sway that we want to solve. Not sure that adding it here is a good idea.

Comment (emersion) [reply kanshi.5.scd]: It's not a bug in sway, it's i3 behaviour that needed to be re-implemented in sway for compatibility. You can just go for `swaymsg 'move workspace to output "Hewlett Packard LA2405 CZ40022966"'` too.

Comment (layus) [reply kanshi.5.scd]: Resolved then ;-)

**Comment by layus** [comment]
Stuck on the CI issue with mixed sign comparison... No yet sure how to fix it.

File: parser.c [hunk 20]
```diff
@@ -338,14 +356,60 @@ static struct kanshi_profile_output *parse_profile_output(
...
+	if (!parser_expect_token(parser, KANSHI_TOKEN_LBRACKET)) {
+		return NULL;
+	}
+
+	// Use the bracket position to generate a default profile name
+	if (profile->name == NULL) {
+		char generated_name[100];
+		int ret = snprintf(generated_name, sizeof(generated_name),
+				"<anonymous at line %d, col %d>", parser->line, parser->col);
+		if (ret >= 0 && ret < sizeof(generated_name)) {
```
Comment (emersion) [inline parser.c]: TBH I'd just drop this check and pick a big enough `generated_name` buffer so that we are sure al `%d`s fit.

Comment (layus) [reply parser.c]: 100 is big enough already. Not combination of two integers can overflow this.

**Comment by layus** [comment]
Okay, left the check out, and force pushed.

**Review by emersion — APPROVED** [review]
Thanks!
